File: 0224-basic-calculator/0224-basic-calculator.cpp
class Solution {
public:
    int calculate(string s) {
        int sum = 0;
        int sign = 1;
        int n = s.size();
        stack<pair<int,int>>stk;
        
        for(int i =0; i<n;i++){
            if(isdigit(s[i])){
                int num = 0;
                while(i<n && isdigit(s[i])){
                    num = num *10 + (s[i] - '0');
                    i++;
                }
                i--;
                sum += (num * sign);
                sign = 1;
            }
            else if(s[i]=='('){
                stk.push({sum, sign});
                sum = 0;
                sign = 1;
            }
            else if(s[i]==')'){
                sum = stk.top().first + (stk.top().second * sum);
                stk.pop();
            }
            else if(s[i]=='-'){
                sign = (-1 * sign);
            }
        }
        return sum;
    }
};

File: 0224-basic-calculator/0224-basic-calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        sum, sign, num, stack = 0, 1, 0, []
        
        for char in s:
            if char.isdigit():
                num = num*10 + int(char)
            elif char in ["+", "-"]:
                sum+= sign * num
                num = 0
                sign =[-1,1][char == "+"]
            elif char == "(" :
                stack.append(sum)
                stack.append(sign)
                sign, sum = 1, 0
            elif char == ")":
                sum+= sign * num
                sum*= stack.pop()
                sum+= stack.pop()
                num = 0
        return sum + num * sign